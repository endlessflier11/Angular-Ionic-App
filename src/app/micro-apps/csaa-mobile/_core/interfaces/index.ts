export * from './auth.interface';
export * from './router';
export * from './http.interface';
export * from './vehicle.interface';
export * from './ionic-lifecycle.interface';
export * from './analytics.interface';
export * from './policy.interface';
export * from './payment.interface';
export * from './config.interface';
export * from './ngxs-store.interface';
