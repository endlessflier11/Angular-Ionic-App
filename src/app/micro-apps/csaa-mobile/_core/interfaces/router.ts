import { Route, UrlTree } from '@angular/router';
import { NavigationOptions } from '@ionic/angular/providers/nav-controller';
import { Observable } from 'rxjs';

interface AppRoute extends Route {
  name?: string;
  isGroupIndex?: boolean; // If true, it means route is the index of parent. Then it inherits all parent properties
}

export type AppRoutes = AppRoute[];

export interface RouteParam {
  [key: string]: string;
}

export interface NamedRouteConfig {
  name: string;
  path: string;
  children?: NamedRouteConfig[];
}

export interface NamedRoute {
  name: string;
  path: string;
  relativePath: string;
}

export interface CanComponentDeactivate {
  canDeactivate: () =>
    | Observable<boolean | UrlTree>
    | Promise<boolean | UrlTree>
    | boolean
    | UrlTree;
}

export interface CsaaRouteStackItem {
  name: string;
  params?: RouteParam;
  options?: NavigationOptions;
}
