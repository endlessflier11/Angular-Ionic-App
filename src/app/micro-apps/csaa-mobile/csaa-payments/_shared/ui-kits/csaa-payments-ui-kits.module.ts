import { NgModule } from '@angular/core';
import { IonicModule } from '@ionic/angular';
import { CommonModule } from '@angular/common';
import { CsaaPaymentHistoryCardComponent } from './csaa-payment-history-card/csaa-payment-history-card.component';
import { UiKitsModule } from '../../../_core/ui-kits/ui-kits.module';
import { CsaaPaymentHistorySelectCardComponent } from './csaa-payment-history-select-card/csaa-payment-history-select-card.component';
import { CsaaPaymentSucceededCardComponent } from './csaa-payment-succeeded-card/csaa-payment-succeeded-card.component';
import { CsaaMakePaymentCardComponent } from './csaa-make-payment-card/csaa-make-payment-card.component';
import { CsaaPayAllPoliciesCardComponent } from './csaa-pay-all-policies-card/csaa-pay-all-policies-card.component';
import { PaymentTermsConditionsComponent } from './payment-terms-conditions/payment-terms-conditions.component';
import { AutopayTermsConditionsComponent } from './autopay-terms-conditions/autopay-terms-conditions.component';
import { CsaaBillsCardComponent } from './csaa-bills-card/csaa-bills-card.component';
import { AutopayDiscountCardComponent } from './autopay-discount-card/autopay-discount-card.component';

const SHARED_COMPONENTS = [
  CsaaPaymentHistoryCardComponent,
  CsaaPaymentHistorySelectCardComponent,
  CsaaMakePaymentCardComponent,
  CsaaPaymentSucceededCardComponent,
  CsaaPayAllPoliciesCardComponent,
  PaymentTermsConditionsComponent,
  AutopayTermsConditionsComponent,
  CsaaBillsCardComponent,
  AutopayDiscountCardComponent,
];

@NgModule({
  declarations: [...SHARED_COMPONENTS],
  imports: [CommonModule, IonicModule, UiKitsModule, IonicModule],
  exports: [...SHARED_COMPONENTS],
})
export class CsaaPaymentsUiKitsModule {}
