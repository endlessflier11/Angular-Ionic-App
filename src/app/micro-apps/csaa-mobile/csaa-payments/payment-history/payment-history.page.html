<ion-header [ngClass]="['csaa', currentTheme]">
  <ion-toolbar>
    <ion-buttons slot="start">
      <csaa-back-button (back)="onClickBackBtn()"></csaa-back-button>
    </ion-buttons>
    <ion-buttons slot="end">
      <ion-button (click)="hoursPopup.show()">Call Service</ion-button>
    </ion-buttons>
  </ion-toolbar>
  <ion-toolbar>
    <ion-title>Payment History</ion-title>
  </ion-toolbar>
</ion-header>
<ion-content [ngClass]="['csaa', currentTheme]">
  <ion-refresher slot="fixed" (ionRefresh)="doRefresh($event.target)">
    <ion-refresher-content></ion-refresher-content>
  </ion-refresher>
  <csaa-payment-history-card
    [history]="(history$ | async)"
    [policy]="(policy$ | async)"
    [loading]="loading"
    headerLine="none"
  >
  </csaa-payment-history-card>
</ion-content>
<csaa-working-hours
  #hoursPopup
  [ngClass]="['csaa', currentTheme]"
  [riskState]="(policy$ | async)?.riskState"
  serviceType="CUSTOMER"
>
</csaa-working-hours>
