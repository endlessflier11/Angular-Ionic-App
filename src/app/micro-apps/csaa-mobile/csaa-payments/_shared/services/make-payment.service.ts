import { Injectable } from '@angular/core';
import { NavigationEnd, Router } from '@angular/router';
import { filter } from 'rxjs/operators';
import { Subscription } from 'rxjs';

import {
  PaymentAccount,
  PaymentResult,
  UpcomingPayment,
  UpcomingPaymentModel,
} from '../../../_core/interfaces';
import { RouterService } from '../../../_core/services';
import { withErrorReporter } from '../../../_core/helpers';
import { CsaaCommonModule } from '../../../csaa-core/csaa-common.module';
import { format } from 'date-fns';

@Injectable({
  providedIn: CsaaCommonModule,
})
export class MakePaymentService {
  private resetListenerSub: Subscription;
  private upcomingPayments: UpcomingPayment[];
  private activePayment: UpcomingPayment;
  private activePaymentAccount: PaymentAccount;
  private paymentResults: PaymentResult = {};

  private returnPathFromAmountMethodPages: { path: string; params?: any };

  public get hasPayment(): boolean {
    return !!this.activePayment;
  }

  constructor(private readonly router: Router, private routerService: RouterService) {}

  startResetListener(): void {
    if (this.resetListenerSub) {return;}
    this.resetListenerSub = this.router.events
      .pipe(filter((e) => e instanceof NavigationEnd))
      .subscribe(
        withErrorReporter((e: NavigationEnd) => {
          if (!e.url.startsWith(this.routerService.fullPath('csaa.payment.index'))) {
            this.reset();
            this.stopResetListener();
          }
        })
      );
  }

  stopResetListener(): void {
    if (!this.resetListenerSub) {return;}
    this.resetListenerSub.unsubscribe();
    this.resetListenerSub = undefined;
  }

  public reset(): void {
    this.setActivePayment(null);
    this.selectPaymentAccount(null);
    this.setReturnPathFromAmountMethodPages(null);
    this.setPaymentResult({});
  }

  public setActivePayment(payment: UpcomingPayment): void {
    this.activePayment = null;
    if (!!payment) {
      this.activePayment = new UpcomingPaymentModel({ ...payment });
    }
    this.startResetListener();
  }

  public getActivePayment() {
    return !!this.activePayment ? new UpcomingPaymentModel({ ...this.activePayment }) : null;
  }

  public updatePaymentAmount(value: number) {
    this.activePayment.otherAmount = value;
  }

  public getPaymentAmount(): number {
    const { amount, otherAmount } = this.activePayment;
    return otherAmount || amount;
  }

  public setUpcomingPayments(payments: UpcomingPayment[]): void {
    this.upcomingPayments = payments;
    this.startResetListener();
  }

  public getUpcomingPayments(): UpcomingPayment[] {
    return this.upcomingPayments;
  }

  public selectPaymentAccount(account: PaymentAccount) {
    this.activePaymentAccount = account;
  }

  public pullSelectedPaymentAccount(): PaymentAccount {
    const activePaymentAccount = this.activePaymentAccount;
    this.activePaymentAccount = null;
    return activePaymentAccount;
  }

  public setReturnPathFromAmountMethodPages(path: string, params = null) {
    this.returnPathFromAmountMethodPages = { path, params };
  }

  public getReturnPathFromAmountMethodPages() {
    return this.returnPathFromAmountMethodPages || { path: 'csaa.payment.index' };
  }

  public setPaymentResult(results: PaymentResult) {
    this.paymentResults = { ...results };
  }

  public pullPaymentResults() {
    const results = this.paymentResults;
    this.paymentResults = null;
    return { ...results };
  }

  public updatePaymentResult(
    policyNumber: string,
    status: string,
    receiptNumber: string,
    paymentAccount: PaymentAccount,
    amount: number
  ) {
    this.paymentResults[policyNumber] = {
      result: status === 'SUCC',
      receiptNumber,
      paymentAccount,
      amount,
      date: format(new Date(Date.now()), 'yyyy-MM-dd'),
    };
  }

  public getPaymentResult(policyNumber: string): boolean {
    return !!this.paymentResults[policyNumber];
  }
}
