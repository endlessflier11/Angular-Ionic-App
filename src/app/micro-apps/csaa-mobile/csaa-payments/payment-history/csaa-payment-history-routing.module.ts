import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AppRoutes } from '../../_core/interfaces';
import { PaymentHistoryPage } from './payment-history.page';

const routes: AppRoutes = [
  {
    isGroupIndex: true,
    path: '',
    component: PaymentHistoryPage,
  },
];

@NgModule({
  declarations: [],
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class CsaaPaymentHistoryRoutingModule {}
