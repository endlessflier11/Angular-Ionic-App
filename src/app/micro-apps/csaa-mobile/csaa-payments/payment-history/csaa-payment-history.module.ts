import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CsaaPaymentHistoryRoutingModule } from './csaa-payment-history-routing.module';
import { IonicModule } from '@ionic/angular';
import { UiKitsModule } from '../../_core/ui-kits/ui-kits.module';
import { CsaaPaymentsUiKitsModule } from '../_shared/ui-kits/csaa-payments-ui-kits.module';

@NgModule({
  imports: [
    CommonModule,
    CsaaPaymentHistoryRoutingModule,
    IonicModule,
    UiKitsModule,
    CsaaPaymentsUiKitsModule,
  ],
})
export class CsaaPaymentHistoryModule {}
