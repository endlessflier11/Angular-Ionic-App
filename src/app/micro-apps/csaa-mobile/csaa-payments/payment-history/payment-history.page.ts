import { Component, OnInit } from '@angular/core';
import { noop, Observable } from 'rxjs';
import { PaymentHistory } from '../../_core/interfaces/payment.interface';
import { finalize, switchMap } from 'rxjs/operators';
import { IonRefresher } from '@ionic/angular';
import { ActivatedRoute } from '@angular/router';
import { withErrorReporter } from '../../_core/helpers';
import { Policy } from '../../_core/interfaces';
import { RouterService } from '../../_core/services';
import { Select, Store } from '@ngxs/store';
import { ConfigState } from '../../_core/store/states/config.state';
import { PaymentState } from '../../_core/store/states/payment.state';
import { CustomerAction, PaymentAction, PolicyAction } from '../../_core/store/actions';
import { PolicyState } from '../../_core/store/states/policy.state';

@Component({
  selector: 'app-csaa-payment-history',
  templateUrl: './payment-history.page.html',
  styleUrls: ['./payment-history.page.scss'],
})
export class PaymentHistoryPage implements OnInit {
  currentTheme: string;
  loading = false;
  private refresher: IonRefresher;
  policyNumber: string;
  policy$: Observable<Policy>;
  @Select(PaymentState.paymentHistory) history$: Observable<PaymentHistory>;

  constructor(
    private readonly store: Store,
    private readonly routerService: RouterService,
    private readonly activatedRoute: ActivatedRoute
  ) {
    this.currentTheme = this.store.selectSnapshot(ConfigState.theme);
  }

  ngOnInit() {
    this.activatedRoute.params.subscribe(
      withErrorReporter((params) => {
        this.policyNumber = params.policyNumber;
        this.initializePage();
      })
    );
  }

  initializePage(): void {
    this.loading = true;
    this.dispatchLoadActions();
    this.policy$ = this.store.select(PolicyState.policyData(this.policyNumber));
  }

  onClickBackBtn() {
    this.routerService.back();
  }

  doRefresh(refresher) {
    this.loading = true;
    this.refresher = refresher;
    this.dispatchReloadActions();
  }

  private completeLoading() {
    this.loading = false;
    if (this.refresher) {
      this.refresher.complete().then(noop);
    }
  }

  private dispatchLoadActions() {
    this.store
      .dispatch(new CustomerAction.LoadCustomer())
      .pipe(
        switchMap(() => this.store.dispatch(new PolicyAction.LoadPolicies())),
        switchMap(() => this.store.dispatch(new PaymentAction.LoadPayments())),
        switchMap(() => this.store.dispatch(new PaymentAction.LoadHistory())),
        finalize(() => this.completeLoading())
      )
      .subscribe(withErrorReporter(noop));
  }

  private dispatchReloadActions() {
    this.store
      .dispatch(new CustomerAction.LoadCustomer())
      .pipe(
        switchMap(() => this.store.dispatch(new PolicyAction.ReloadPolicies())),
        switchMap(() => this.store.dispatch(new PaymentAction.ReloadPayments())),
        switchMap(() => this.store.dispatch(new PaymentAction.ReloadHistory())),
        finalize(() => this.completeLoading())
      )
      .subscribe(withErrorReporter(noop));
  }
}
