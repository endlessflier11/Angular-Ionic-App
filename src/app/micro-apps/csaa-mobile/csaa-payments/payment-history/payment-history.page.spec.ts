import { ComponentFixture, fakeAsync, TestBed } from '@angular/core/testing';

import { PaymentHistoryPage } from './payment-history.page';
import {
  BillingHistoryResponseMock,
  BillingSummaryResponseMock,
  By,
  clickEvent,
  CustomerResponseMock,
  PageTestingModule,
  PoliciesResponseMock,
  StoreTestBuilder,
  WalletResponseMock,
} from '@app/testing';
import { UiKitsModule } from '../../_core/ui-kits/ui-kits.module';
import { CsaaPaymentsUiKitsModule } from '../_shared/ui-kits/csaa-payments-ui-kits.module';
import { ActivatedRoute } from '@angular/router';
import { of } from 'rxjs';
import { CallService } from '../../_core/services/call.service';
import { CsaaAppInjector } from '../../csaa-app.injector';
import { Injector } from '@angular/core';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { CsaaHttpClientModule } from '../../_core/csaa-http-client.module';
import { AnalyticsService } from '../../_core/services';
import { CsaaCoreModule } from '../../csaa-core/csaa-core.module';
import { Category, EventName } from '../../_core/interfaces';
import { Store } from '@ngxs/store';
import { Platform } from '@ionic/angular';

describe('PaymentHistoryPage', () => {
  let component: PaymentHistoryPage;
  let fixture: ComponentFixture<PaymentHistoryPage>;
  let httpTestingController: HttpTestingController;
  let analyticsService;

  beforeEach(async () => {
    try {
      await TestBed.configureTestingModule({
        declarations: [PaymentHistoryPage],
        imports: [
          PageTestingModule.withConfig({
            providesAnalytics: true,
            providesConfig: true,
            providesStorage: true,
          }),
          UiKitsModule,
          CsaaPaymentsUiKitsModule,
          CsaaCoreModule.forRoot(),
        ],
        providers: [
          HttpClientTestingModule,
          CsaaHttpClientModule,
          { provide: ActivatedRoute, useValue: { params: of({ policyNumber: 'demo' }) } },
          { provide: CallService, useValue: { call: jest.fn() } },
        ],
      }).compileComponents();

      CsaaAppInjector.injector = TestBed.inject(Injector);
      const store = TestBed.inject(Store);
      StoreTestBuilder.withDefaultMocks().resetStateOf(store);
      await TestBed.inject(Platform).ready();

      analyticsService = TestBed.inject(AnalyticsService);
      httpTestingController = TestBed.inject(HttpTestingController);
      fixture = TestBed.createComponent(PaymentHistoryPage);
      component = fixture.componentInstance;
      fixture.detectChanges();
    } catch (error) {
      console.error(error);
      fail(error.message);
    }
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('Registration ID is Null', () => {
    const MOCK: { [k: string]: any } = {};
    let flushRequestsInGroup = (_?) => {};

    beforeEach(fakeAsync(() => {
      // Initialize component data
      MOCK.customer = CustomerResponseMock.create().toJson();
      MOCK.policies = [PoliciesResponseMock.create().toJson()[0]];
      MOCK.billingHistory = BillingHistoryResponseMock.create().toJson();
      MOCK.billingSummary = BillingSummaryResponseMock.create().withPaymentDue().toJson();
      MOCK.wallet = WalletResponseMock.create().toJson();

      // const [registration] = MOCK.customer.registrations;
      // const ownerId = registration.registrationId;

      flushRequestsInGroup = (_MOCK = MOCK) => {
        // httpTestingController
        //   .match(AppEndpointsEnum[AppEndpointsEnum.customerSearch])
        //   .forEach((req) => req.flush(_MOCK.customer));
        // httpTestingController
        //   .match(AppEndpointsEnum[AppEndpointsEnum.policies])
        //   .forEach((req) => req.flush(_MOCK.policies));
        // httpTestingController
        //   .match(AppEndpointsEnum[AppEndpointsEnum.billingSummary])
        //   .forEach((req) => req.flush(_MOCK.billingSummary));
        // httpTestingController
        //   .match(AppEndpointsEnum[AppEndpointsEnum.customerSearch])
        //   .forEach((req) => req.flush(_MOCK.customer));
        // httpTestingController
        //   .match(AppEndpointsEnum[AppEndpointsEnum.policies])
        //   .forEach((req) => req.flush(_MOCK.policies));
        // httpTestingController
        //   .match(AppEndpointsEnum[AppEndpointsEnum.billingSummary])
        //   .forEach((req) => req.flush(_MOCK.billingSummary));
        //
        // httpTestingController
        //   .match(AppEndpointsEnum[AppEndpointsEnum.billingHistory])
        //   .pop()
        //   .flush(MOCK.billingHistory);
        //
        // httpTestingController
        //   .match(`${AppEndpointsEnum[AppEndpointsEnum.billingWallet]}?ownerId=${ownerId}`)
        //   .forEach((req) => req.flush(WalletResponseMock.create().toJson()));
      };
    }));

    afterEach(() => {
      // Ensures no pending request
      httpTestingController.verify();
    });

    it('should match snapshot when registration ID is null', () => {
      component.ngOnInit();
      flushRequestsInGroup();
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });

    it('should track the call button', () => {
      component.ngOnInit();
      flushRequestsInGroup();
      fixture.detectChanges();

      const callBtn = fixture.debugElement.query(
        By.cssAndText('csaa-working-hours ion-button', 'Call Service')
      );
      callBtn.triggerEventHandler('click', clickEvent);
      expect(analyticsService.trackEvent).toHaveBeenCalledWith(
        EventName.CONTACT_INITIATED,
        Category.global,
        {
          event_type: 'Option Selected',
          selection: 'Call Service',
          method: 'phone',
        }
      );
    });
  });
});
