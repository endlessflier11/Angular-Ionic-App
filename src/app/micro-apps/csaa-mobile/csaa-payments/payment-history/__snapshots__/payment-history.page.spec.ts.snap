// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentHistoryPage Registration ID is Null should match snapshot when registration ID is null 1`] = `
<app-csaa-payment-history
  __history$__selector={[Function Observable]}
  __ngContext__={[Function LRootView]}
  activatedRoute={[Function Object]}
  currentTheme=""
  loading="false"
  policy$={[Function Observable]}
  policyNumber={[Function String]}
  routerService={[Function RouterService]}
  store={[Function Store]}
>
  <ion-header
    class="csaa"
  >
    <ion-toolbar>
      <ion-buttons
        slot="start"
      >
        <csaa-back-button>
          <ion-button>
            <ion-icon
              name="chevron-back-outline"
              slot="start"
            />
             Back
          </ion-button>
        </csaa-back-button>
      </ion-buttons>
      <ion-buttons
        slot="end"
      >
        <ion-button>
          Call Service
        </ion-button>
      </ion-buttons>
    </ion-toolbar>
    <ion-toolbar>
      <ion-title>
        Payment History
      </ion-title>
    </ion-toolbar>
  </ion-header><ion-content
    class="csaa"
  >
    <ion-refresher
      slot="fixed"
    >
      <ion-refresher-content />
    </ion-refresher>
    <csaa-payment-history-card
      headerline="none"
    >
      <ion-card
        class="csaa-card"
      >
        <ion-card-content
          class="csaa-item-list"
        >
          <ion-list>
          </ion-list>
        </ion-card-content>
      </ion-card>
    </csaa-payment-history-card>
  </ion-content><csaa-working-hours
    class="csaa"
    servicetype="CUSTOMER"
  >
    <div
      class="csaa-working-hours hidden"
    >
      <ion-card
        class="card card-md"
      >
        <ion-card-header>
          Service Hours
        </ion-card-header>
        <ion-card-content
          class="card-content card-content-md"
        >
          <ion-list>
            <ion-item>
            </ion-item>
          </ion-list>
          <ion-button
            expand="full"
          >
             Call Service 
          </ion-button>
        </ion-card-content>
      </ion-card>
    </div>
  </csaa-working-hours>
</app-csaa-payment-history>
`;
